### PR TITLE
PMM-8459 Ability to disable updates.

### PR DESCRIPTION
**Pull request by JiriCtvrtka**
[PMM-8459](https://jira.percona.com/browse/PMM-8459)

- [x] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1939

- [x] Links to other linked pull requests (optional).
pmm: https://github.com/percona/pmm/pull/771
api-tests: https://github.com/Percona-Lab/pmm-api-tests/pull/228
---
- [x] Tests passed.
- [ ] Feature build pass.
- [x] (Re)requested review.
- [x] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

### CODE REVIEW

**Comment by JiriCtvrtka** [comment]
Updates can be disabled only when telemetry is disabled too.

File: services/server/server.go [hunk 1]
```diff
@@ -522,6 +522,13 @@ func (s *Server) validateChangeSettingsRequest(ctx context.Context, req *serverp
 
 	// check request parameters compatibility with environment variables
 
+	if req.DisableUpdates && s.envSettings.EnableUpdates {
+		return status.Error(codes.FailedPrecondition, "Updates are enabled via ENABLE_UPDATES environment variable.")
+	}
+	if req.DisableUpdates && s.envSettings.EnableTelemetry {
+		return status.Error(codes.FailedPrecondition, "Updates cannot be disabled because telemetry is enabled via ENABLE_TELEMETRY environment variable.")
+	}
+
```
Comment (BupycHuk) [inline services/server/server.go:531]: user also should not be able to enable updates if it was disabled via env variable.

File: services/server/server.go [hunk 2]
```diff
@@ -429,7 +429,7 @@ func (s *Server) readUpdateAuthToken() (string, error) {
 // convertSettings merges database settings and settings from environment variables into API response.
 func (s *Server) convertSettings(settings *models.Settings) *serverpb.Settings {
 	res := &serverpb.Settings{
-		UpdatesDisabled:  s.envSettings.DisableUpdates,
+		UpdatesDisabled:  !settings.Updates.Disabled,
```
Comment (BupycHuk) [inline services/server/server.go:432]:
```suggestion
		UpdatesDisabled:  settings.Updates.Disabled,
```

File: services/server/server_test.go [hunk 3]
```diff
@@ -188,8 +188,22 @@ func TestServer(t *testing.T) {
 			RemoveAlertManagerRules: true,
 		}))
 
-		s.envSettings.DisableTelemetry = true
+		s.envSettings.EnableUpdates = true
+		expected = status.New(codes.FailedPrecondition, "Updates are enabled via ENABLE_UPDATES environment variable.")
```
Comment (BupycHuk) [inline services/server/server_test.go:192]: Do we have `ENABLE_UPDATES` env variable? As far as I know we have only DISABLE_UPDATES

Comment (BupycHuk) [reply services/server/server_test.go:192]: let's rewrite test taking into account that we can't enable updates if it's disabled.

File: services/server/server.go [hunk 4]
```diff
@@ -522,6 +522,13 @@ func (s *Server) validateChangeSettingsRequest(ctx context.Context, req *serverp
 
 	// check request parameters compatibility with environment variables
 
+	if req.DisableUpdates && s.envSettings.EnableUpdates {
+		return status.Error(codes.FailedPrecondition, "Updates are enabled via ENABLE_UPDATES environment variable.")
+	}
+	if req.DisableUpdates && s.envSettings.EnableTelemetry {
+		return status.Error(codes.FailedPrecondition, "Updates cannot be disabled because telemetry is enabled via ENABLE_TELEMETRY environment variable.")
+	}
```
Comment (BupycHuk) [inline services/server/server.go:530]: updates and telemetry is independent, so we can remove this check.

File: models/settings_helpers.go [hunk 5]
```diff
@@ -422,9 +428,14 @@ func validateSettingsConflicts(params *ChangeSettingsParams, settings *Settings)
...
 		return fmt.Errorf("Cannot disable telemetry while STT is enabled.") //nolint:golint,stylecheck
 	}
-
 	if params.LogOut && (params.Email != "" || params.SessionID != "") {
-		return fmt.Errorf("Cannot loguot while updating Percona Platform user data.") //nolint:golint,stylecheck
+		return fmt.Errorf("Cannot logout while updating Percona Platform user data.") //nolint:golint,stylecheck
+	}
+	if params.DisableUpdates && !params.DisableTelemetry && !settings.Telemetry.Disabled {
+		return fmt.Errorf("Cannot disable updates while telemetry is enabled.") //nolint:golint,stylecheck
+	}
```
Comment (BupycHuk) [inline models/settings_helpers.go:436]: let's remove this check.

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Do not forget about API tests.

File: services/server/server.go (hunk 4 above)
Comment (JiriCtvrtka) [reply services/server/server.go:530]: Here https://perconacorp.slack.com/archives/C03J8FZFU/p1626470324102700 is discussion that they are coupled. Ok I will change everything related to it.

Comment (BupycHuk) [reply services/server/server.go]: Please discuss it with Roma, Steve or Denys. But previously users were able to disable updates using env variable without disabling telemetry.

File: services/server/server.go [hunk 6]
```diff
@@ -522,6 +522,13 @@ func (s *Server) validateChangeSettingsRequest(ctx context.Context, req *serverp
 
 	// check request parameters compatibility with environment variables
 
+	if req.DisableUpdates && !s.envSettings.DisableUpdates {
+		return status.Error(codes.FailedPrecondition, "Updates are enabled via DISABLE_UPDATES environment variable.")
+	}
```
Comment (BupycHuk) [inline services/server/server.go:527]: let's remove it. `s.envSettings.DisableUpdates` is false by default, so users won't be able to disable updates.

File: services/server/server.go (hunk 4 above)
Comment (denisok) [reply services/server/server.go]: that was just assumption from Steve (that disable doesn't work because of telemetry). It is not a requirement.

Comment (JiriCtvrtka) [reply services/server/server.go]: @denisok Ok thanks for verifying.

**Review by percona-csalguero — APPROVED** [review]
LGTM, just address the changes requested by Nurlan.

**Comment by JiriCtvrtka** [comment]
@BupycHuk Take a look if I got you right. API tests added in description.

**Review by askomorokhov — APPROVED** [review]

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**